Issue 8

trikey opened the pull request.

trikey committed:
Add answers
Add test passages

trikey committed:
Rubocop fix

trikey committed:
Add number of question, added percentage of result

trikey committed:
Rubocop fix

psylone commented on app/models/test_passage.rb:
```diff
@@ -0,0 +1,39 @@
+
+  def next_question
+    self.current_question_number += 1
+    test.questions.order(:id).where('id > ?', current_question.id).first
```
Исходя из названия метода не ясно что он изменяет значение какой-то переменной. Но даже если помеять название, будет лучше обойтись без этого счётчика. Поддерживать его будет довольно неудобно.

psylone commented on app/models/test_passage.rb:
```diff
@@ -0,0 +1,39 @@
+
+  def accept!(answer_id)
+    self.correct_questions += 1 if correct_answer?(answer_id)
+    self.current_question = next_question
```
По заданию необходимо устанавливать следующий вопрос с помощью обратного вызова.

psylone commented on app/views/test_passages/result.html.erb:
```diff
@@ -0,0 +1,8 @@
+<h1>The <%= @test_passage.test.title %> completed!</h1>
+<p><%= @percent %>%</p>
+<% if @percent >= 85 %>
```
Если посмотреть на код представления, неясно что за переменная `@percent`, к какому контексту она относится.

psylone commented on app/controllers/test_passages_controller.rb:
```diff
@@ -0,0 +1,25 @@
+  def show; end
+
+  def result
+    @percent = (@test_passage.correct_questions * 100 / @test_passage.test.questions.count).round
```
Это должно быть либо в модели, либо в хэлпере. Ведь мы можем заохотеть выводить эти же данные где-то ещё, например, в письме которое будем отправлять пользователю по завершению теста. Или в списке всех пройденных тестов.

psylone commented on app/views/test_passages/show.html.erb:
```diff
@@ -0,0 +1,20 @@
+<h1>Pass the <%= @test_passage.test.title %>></h1>
+
+<p>Всего вопросов: <%= @test_passage.test.questions.count %></p>
+<p>Вопрос №<%= @test_passage.current_question_number %></p>
```
Понимаю логику такого манёвра, и на первый взгляд, решение кажется подходящим. Но отдельную переменную с индексом вопроса трудно поддерживать, о ней постоянно нужно помнить когда мы совершаем какие-то действия с объектом `TestPassage`.

Будет лучше либо как-то сортировать вопросы, либо в таблицу вопросов добавить поле с позицией вопроса. Тогда мы, во-первых, сможем отсортировать по этому полю, а во-вторых, изменять значение этого поля из панели администратора.

psylone commented on app/controllers/answers_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  before_action :find_question, only: %i[create new]
+  before_action :find_answer, only: %i[show edit update destroy]
+
+  rescue_from ActiveRecord::RecordNotFound, with: :rescue_with_answer_not_found
```
Эта штука общего характера, её стоит вынести выше в базовый контроллер.

psylone commented on app/models/test_passage.rb:
```diff
@@ -0,0 +1,39 @@
+class TestPassage < ApplicationRecord
+  PASSING = 1
+  PASSED = 3
+  PASSED_WITH_ERRORS = 2
```
Посмотри в сторону модуля [`enum`](http://api.rubyonrails.org/classes/ActiveRecord/Enum.html#method-i-enum).

trikey committed:
Fix bugs
Move correct answers percentage count to model
Added question number for ordering questions in test
404 Exception is now processed in application_controller

psylone commented on app/controllers/test_passages_controller.rb:
```diff
@@ -4,7 +4,7 @@ class TestPassagesController < ApplicationController
 
   def result
-    @percent = (@test_passage.correct_questions * 100 / @test_passage.test.questions.count).round
+    @correct_answers_percent = @test_passage.correct_answers_percent
```
Ну, и раз у нас `@test_passage` доступна из представлений, нет смысла здесь отдельную инстанс-переменную заводить.

psylone commented on db/schema.rb:
```diff
@@ -32,6 +32,7 @@
     t.bigint "test_id", null: false
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.integer "number"
```
Стоит также обновить форму создания/редактирования вопроса чтобы задавать его порядковый номер в тесте.

psylone commented on app/models/test_passage.rb:
```diff
@@ -1,39 +1,39 @@
   end
 
   def before_validation_set_first_question
     self.current_question = test.questions.first
```
Здесь поскольку у нас появился порядковый номер вопроса, необходимо сортировать по этому атрибуту.

psylone commented on app/models/test_passage.rb:
```diff
@@ -1,39 +1,39 @@
+  scope :passing, -> { where(status: status[:passing]) }
 
   before_validation :before_validation_set_first_question, on: :create
+  before_update :next_question, on: :update
```
Эти 2 обратных вызова можно объединить в один, поскольку по логике они делают одно и то же.

trikey committed:
Fix bugs
